companion/src/firmwares/moduledata.cpp

```diff
@@ -262,3 +268,274 @@ int ModuleData::getMaxChannelCount()
+
+                          { PULSES_OFF, MODULE_TYPE_NONE },
+                          { PULSES_PPM, MODULE_TYPE_PPM },
+
+                          { PULSES_PXX_XJT_X16, MODULE_TYPE_XJT_PXX1 },
+                          { PULSES_PXX_XJT_D8, MODULE_TYPE_XJT_PXX1 },
+                          { PULSES_PXX_XJT_LR12, MODULE_TYPE_XJT_PXX1 },
+
+                          { PULSES_ACCESS_ISRM, MODULE_TYPE_ISRM_PXX2 },
+                          { PULSES_ACCST_ISRM_D16, MODULE_TYPE_ISRM_PXX2 },
+
+                          { PULSES_LP45, MODULE_TYPE_DSM2 },
+                          { PULSES_DSM2, MODULE_TYPE_DSM2 },
+                          { PULSES_DSMX, MODULE_TYPE_DSM2 },
+
+                          { PULSES_CROSSFIRE, MODULE_TYPE_CROSSFIRE },
+                          { PULSES_MULTIMODULE, MODULE_TYPE_MULTIMODULE },
+                          { PULSES_PXX_R9M, MODULE_TYPE_R9M_PXX1 },
+                          { PULSES_ACCESS_R9M, MODULE_TYPE_R9M_PXX2 },
+                          { PULSES_PXX_R9M_LITE, MODULE_TYPE_R9M_LITE_PXX1 },
+                          { PULSES_ACCESS_R9M_LITE, MODULE_TYPE_R9M_LITE_PXX2 },
+                          { PULSES_GHOST, MODULE_TYPE_GHOST },
+                          { PULSES_ACCESS_R9M_LITE_PRO, MODULE_TYPE_R9M_LITE_PRO_PXX2 },
+                          { PULSES_SBUS, MODULE_TYPE_SBUS },
+
+                          { PULSES_XJT_LITE_X16, MODULE_TYPE_XJT_LITE_PXX2 },
+                          { PULSES_XJT_LITE_D8, MODULE_TYPE_XJT_LITE_PXX2 },
+                          { PULSES_XJT_LITE_LR12, MODULE_TYPE_XJT_LITE_PXX2 },
+
+                          { PULSES_AFHDS3, MODULE_TYPE_FLYSKY }
+                      };
+
+  QPair<int, int>elmt;
+
+  foreach (elmt, ProtocolTypeTable) {
+    if (elmt.first == (int)protocol)
+      return elmt.second;
+  }
+
+  return -1;
+}
+
+//  static
+int ModuleData::getSubTypeFromProtocol(unsigned int protocol)
+{
+  if (protocol == PULSES_ACCST_ISRM_D16 || protocol == PULSES_ACCESS_ISRM)
+    return protocol - PULSES_ACCESS_ISRM;
+  else if (protocol >= PULSES_PXX_XJT_X16 && protocol <= PULSES_PXX_XJT_LR12)
+    return protocol - PULSES_PXX_XJT_X16;
+  else
+    return 0;
+}
+
+//  static
+QString ModuleData::typeToString(int type)
+{
+  static const char * strings[] = {
+    "OFF",
+    "PPM",
+    "FrSky XJT",
+    "ISRM",
+    "DSM2",
+    "TBS Crossfire",
+    "DIY Multiprotocol Module",
+    "FrSky R9M",
+    "FrSky R9M ACCESS",
+    "FrSky R9MLite",
+    "FrSky R9ML ACCESS",
+    "ImmersionRC Ghost",
+    "FrSky R9MLP ACCESS",
+    "SBUS output at VBat",
+    "FrSky XJT Lite",
+    "FlySky"
+  };
+
+  return CHECK_IN_ARRAY(strings, type);
+}
+
+//  static
+bool ModuleData::isInternalModuleAvailable(int moduleType)
+{
+  if (moduleType == MODULE_TYPE_NONE)
+    return true;
+
+  Firmware *fw = getCurrentFirmware();
+  Board::Type board = fw->getBoard();
+  QString id = fw->getId();
+
+  if (!Boards::getCapability(board, Board::HasInternalModuleSupport))
```

Comment:
Whereby the setting is not "configurable", these radios still have an internal module (which is fixed).